src/wodan/keyedmap.ml

```diff
@@ -0,0 +1,113 @@
+open Stdcompat
+
+module type OrderedType = sig
+  type t
+
+  val compare : t -> t -> int
+end
+
+module Make (Ord : OrderedType) = struct
+  exception Already_exists
+
+  module M = Map.Make (Ord)
+
+  type 'a t = 'a M.t ref
+
+  type key = Ord.t
+
+  let create () = ref M.empty
+
+  let length m = M.cardinal !m
+
+  let clear m = m := M.empty
+
+  let find m k = M.find k !m
+
+  let find_opt m k = M.find_opt k !m
+
+  let mem m k = M.mem k !m
+
+  let add m k v = m := M.add k v !m
+
+  let update m k f = m := M.update k f !m
+
+  let replace_if_present m k v =
+    m :=
+      M.update k
+        (function
+          | Some _ ->
+              Some v
+          | None ->
+              raise Not_found)
+        !m
+
+  let add_if_absent m k v =
```

Comment:
Same as above.
I prefer the name xadd.
"if" seems to imply nothing happens in the other case.